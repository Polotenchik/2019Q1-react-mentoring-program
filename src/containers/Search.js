import React from 'react';
import { Input } from '../common/Input';
import { Title } from '../common/Title';
import { Button } from '../common/Button';
import { Label } from '../common/Label';

export const Search = () => {
    return  (
        <div className={'header-search'}>
            <Title content='Find your movie' type='headline' />
            <Input />
            <div className={'search-components'}>
                <div className={'search-type-section'}>
                    <Label content={'search by'} type='search' />
                    <Button content='Title' type='search' />
                    <Button content='Genre' type='search' />
                </div>
                <div>
                    <Button content='Search' type='search' />
                </div>
            </div>
        </div>
    );
};