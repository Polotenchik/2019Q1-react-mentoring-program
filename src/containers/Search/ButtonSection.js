import React from 'react';
import { Button } from  '../../common';

export const ButtonSection = ({searchButtonClick}) => {
    return (
        <Button 
            content='Search' 
            btnClick={ searchButtonClick } 
            btnRed 
            btnBig
        />
    );
}