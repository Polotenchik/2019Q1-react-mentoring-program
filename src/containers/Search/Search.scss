.header-search {
    margin-top: 100px;
}

.search-components {
    display: flex;
    margin-top: 10px;
    justify-content: space-between;
    align-items: center;
}

.search-type-section {
    display: flex;
}
  
.search-type-section > div:nth-child(2n){
    margin: 0 $elementPadding;
}