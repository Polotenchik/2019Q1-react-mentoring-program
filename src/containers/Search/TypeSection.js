import React from 'react';
import { Button, Label } from '../../common';

const TypeSection = ({ searchTypes, searchTypeClick }) => (
    <div className={'search-type-section'}>
        <Label content={'search by'} labelUC labelWhite />
        { searchTypes.list.map(item => (
                <Button
                    btnRed={ item === searchTypes.active }
                    btnGrey={ item !== searchTypes.active }
                    key={ item }
                    content={ item }
                    btnClick={ searchTypeClick }
                    btnInline
                    btnUC
                />
        ))
        }
    </div>
);

export default TypeSection;
