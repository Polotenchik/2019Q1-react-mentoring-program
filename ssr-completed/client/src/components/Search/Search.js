import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import Logo from '../../common/Components/Logo/Logo';
import RadioButton from '../../common/Components/RadioButton/RadioButton';
import Button from '../../common/Components/Button/Button';
import Theme from '../../common/Theme/Theme';
import HeaderBackground from '../../common/Components/HeaderBackground/HeaderBackground';
import HeaderWrapper from '../../common/Components/HeaderWrapper/HeaderWrapper';

const Search = ({
  searchBy,
  inputChangeHandler,
  toggleSearchBy,
  formSubmitHandler,
}) => {
  return (
    <HeaderBackground>
      <HeaderWrapper>
        <Logo />
        <Title main>find your movie</Title>
        <Form onSubmit={formSubmitHandler}>
          <InputWrapper>
            <Input
              id="searchInput"
              placeholder="Search here..."
              onChange={(evt) => inputChangeHandler(evt)}
              required
              autoFocus />
            <LabelArrow htmlFor="searchInput">
              <i className="fas fa-level-down-alt" />
            </LabelArrow>
          </InputWrapper>
          <SearchWrapper>
            <Title small>search by</Title>
            <RadioButton
              inputName="searchBy"
              inputValue="title"
              inputId="title"
              filtered={searchBy}
              handlerChange={toggleSearchBy}
              colored />
            <RadioButton
              inputName="searchBy"
              inputValue="genre"
              inputId="genres"
              filtered={searchBy}
              handlerChange={toggleSearchBy}
              colored />
            <Button id="searchBtn" type="submit" red>search</Button>
          </SearchWrapper>
        </Form>
      </HeaderWrapper>
    </HeaderBackground>
  );
};

export default Search;

Search.propTypes = {
  searchBy: PropTypes.string.isRequired,
  inputChangeHandler: PropTypes.func.isRequired,
  toggleSearchBy: PropTypes.func.isRequired,
  formSubmitHandler: PropTypes.func.isRequired,
};

const Title = styled.h2`
    color: ${Theme.colors.white};    
    text-transform: uppercase;
    
    
    ${props => props.main && css`
        font-size: 20px;
        font-weight: 800;
    `}
    
    ${props => props.small && css`
        font-size: 16px;
    `}
`;

const Form = styled.form`
    width: 100%;    
`;

const InputWrapper = styled.div`
    width: 100%;
    position: relative;
`;

const Input = styled.input`
    width: 100%;  
    box-sizing: border-box;  
    padding: 10px;
    background-color: ${Theme.colors.sub_grey};;
    color: ${Theme.colors.white};
    border: 0;
    border-bottom: 2px solid ${Theme.colors.red};
    outline: none;
`;

const LabelArrow = styled.label`
    position: absolute;
    top: 10px;
    right: 10px;
    color: ${Theme.colors.medium_grey};
    transform: rotate(90deg);  
`;

const SearchWrapper = styled.div.attrs({
  className: 'search-wrapper',
})`
    display: flex;
    align-items: center;  
`;
