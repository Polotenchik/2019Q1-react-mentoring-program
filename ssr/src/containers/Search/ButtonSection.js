import React from 'react';
import { Link, Route } from 'react-router-dom';
import { Button } from '../../common';

export const ButtonSection = ({ searchButtonClick }) => (
        <Link to="/search">
            <Button
                content='Search'
                btnClick={ searchButtonClick }
                btnRed
                btnBig
            />
        </Link>
);
