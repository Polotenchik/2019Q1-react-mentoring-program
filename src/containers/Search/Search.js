import React from 'react';
import { connect } from 'react-redux';
import { startSearch } from '../../redux/search/search.actions';
import { Input, Title } from '../../common';
import { TypeSection } from './TypeSection';
import { ButtonSection } from './ButtonSection';

class Search extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
          searchTypes: {
            list: ['title', 'genres'],
            active: 'title'
          },
          currentInputValue: ''
        };
    }

    startSearching = () => {
        this.props.onSearch(this.state.currentInputValue, this.state.searchTypes.active);
        this.setState({ currentInputValue: '' })
    };

    enterKeyPressedOnInput = (e) => {
        if (e.key === 'Enter') {
          this.startSearching();
        }
    };

    changeSearchType = (newType) => {
        this.setState({
            searchTypes: {
                list: this.state.searchTypes.list, 
                active: newType 
            }
        });
    };

    cloneSearchInput = (el) => {
        this.setState({currentInputValue: el.target.value});
    };

    render() {
        const { searchMode } = this.props;

        if (searchMode) {
            return  (
                <div className={'header-search'}>
                    <Title content='Find your movie' ttlWhite ttlUC />
                    <Input 
                        onStartTyping={ this.cloneSearchInput }
                        value={ this.state.currentInputValue }
                        onKeyEnterPressed={ this.enterKeyPressedOnInput } 
                    />
                    <div className={'search-components'}>
                        <TypeSection 
                            searchTypes={ this.state.searchTypes }
                            searchTypeClick={ this.changeSearchType }
                        />
                        <ButtonSection 
                            parentFormId={ 'search-form' }
                            searchButtonClick={ this.startSearching }
                        />
                    </div>
                </div>
            );
        } else {
            return null;
        }
    }
};

const mapStateToProps = store => ({
    searchMode: store.mode.search
});

const mapDispatchToProps = dispatch => ({
    onSearch(phrase, type) {
        dispatch(startSearch(phrase, type))
    }
});
  
export default connect(mapStateToProps, mapDispatchToProps)(Search);