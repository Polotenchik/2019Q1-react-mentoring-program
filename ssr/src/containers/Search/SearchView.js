import React from 'react';
import { Input, Title } from '../../common';
import { TypeSection } from './TypeSection';
import { ButtonSection } from './ButtonSection';

export default class SearchView extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      searchTypes: {
        list: ['title', 'genres'],
        active: this.props.searchType,
      },
      currentInputValue: '',
    };
  }

    startSearching = () => {
      this.props.onSearch(this.state.currentInputValue, this.state.searchTypes.active);
      this.setState({ currentInputValue: this.state.search.phrase });
    };

    enterKeyPressedOnInput = (e) => {
      if (e.key === 'Enter') {
        this.startSearching();
      }
    };

    changeSearchType = (newType) => {
      this.setState({
        searchTypes: {
          list: this.state.searchTypes.list,
          active: newType,
        },
      });
    };

    cloneSearchInput = (el) => {
      this.setState({ currentInputValue: el.target.value });
    };

    render() {
      const { searchMode } = this.props;

      if (searchMode) {
        return (
                <div className={'header-search'}>
                    <Title content='Find your movie' ttlWhite ttlUC />
                    <Input
                        onStartTyping={ this.cloneSearchInput }
                        value={ this.state.currentInputValue }
                        onKeyEnterPressed={ this.enterKeyPressedOnInput }
                    />
                    <div className={'search-components'}>
                        <TypeSection
                            searchTypes={ this.state.searchTypes }
                            searchTypeClick={ this.changeSearchType }
                        />
                        <ButtonSection
                            parentFormId={ 'search-form' }
                            searchButtonClick={ this.startSearching }
                        />
                    </div>
                </div>
        );
      }
      return null;
    }
}
