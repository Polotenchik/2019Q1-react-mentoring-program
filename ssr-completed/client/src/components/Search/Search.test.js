import React from 'react';
import { shallow } from 'enzyme';
import Search from './Search';

describe('Search Component Snapshot', () => {
    it('Search Component should render correctly with given props', () => {
        const searchBy='title';
        const inputChangeHandler=jest.fn();
        const toggleSearchBy=jest.fn();
        const formSubmitHandler=jest.fn();
        const component = shallow(<Search searchBy={searchBy} inputChangeHandler={inputChangeHandler} toggleSearchBy={toggleSearchBy} formSubmitHandler={formSubmitHandler}/>);
        expect(component).toMatchSnapshot();
    });
});

