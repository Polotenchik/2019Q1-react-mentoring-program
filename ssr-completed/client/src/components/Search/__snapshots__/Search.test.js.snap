// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component Snapshot Search Component should render correctly with given props 1`] = `
<styled.div>
  <styled.div>
    <Logo />
    <styled.h2
      main={true}
    >
      find your movie
    </styled.h2>
    <styled.form
      onSubmit={[MockFunction]}
    >
      <styled.div>
        <styled.input
          autoFocus={true}
          id="searchInput"
          onChange={[Function]}
          placeholder="Search here..."
          required={true}
        />
        <styled.label
          htmlFor="searchInput"
        >
          <i
            className="fas fa-level-down-alt"
          />
        </styled.label>
      </styled.div>
      <styled.div>
        <styled.h2
          small={true}
        >
          search by
        </styled.h2>
        <RadioButton
          colored={true}
          filtered="title"
          handlerChange={[MockFunction]}
          inputId="title"
          inputName="searchBy"
          inputValue="title"
        />
        <RadioButton
          colored={true}
          filtered="title"
          handlerChange={[MockFunction]}
          inputId="genres"
          inputName="searchBy"
          inputValue="genre"
        />
        <styled.button
          id="searchBtn"
          red={true}
          type="submit"
        >
          search
        </styled.button>
      </styled.div>
    </styled.form>
  </styled.div>
</styled.div>
`;
